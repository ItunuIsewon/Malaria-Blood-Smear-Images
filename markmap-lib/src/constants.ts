export const template = process.env.TEMPLATE || '';

export const baseJsPaths = [
  `d3@${process.env.D3_VERSION}/dist/d3.min.js`,
  `markmap-view@${process.env.VIEW_VERSION}/dist/browser/index.js`,
];
