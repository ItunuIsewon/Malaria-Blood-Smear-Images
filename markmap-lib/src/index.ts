export * from './types';
export * from './constants';
export * from './transform';

export const transformerVersions = {
  'markmap-lib': 'process.env.VERSION',
  d3: process.env.D3_VERSION,
};
