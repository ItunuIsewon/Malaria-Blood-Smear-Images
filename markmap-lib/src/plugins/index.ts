import katex from './katex';
import prism from './prism';
import frontmatter from './frontmatter';
import npmUrl from './npm-url';
import hljs from './hljs';

export * from './base';
export const plugins = [frontmatter, katex, hljs, npmUrl];
