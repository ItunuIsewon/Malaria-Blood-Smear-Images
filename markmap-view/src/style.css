.markmap {
  font: 300 16px/20px sans-serif;

  &-link {
    fill: none;
  }

  &-node {
    > circle {
      cursor: pointer;
    }
  }

  &-foreign {
    display: inline-block;

    a {
      color: #0097e6;
      &:hover {
        color: #00a8ff;
      }
    }

    code {
      padding: 0.25em;
      font-size: calc(1em - 2px);
      color: #555;
      background-color: #f0f0f0;
      border-radius: 2px;
    }

    pre {
      margin: 0;
      > code {
        display: block;
      }
    }

    del {
      text-decoration: line-through;
    }

    em {
      font-style: italic;
    }

    strong {
      font-weight: bold;
    }

    mark {
      background: #ffeaa7;
    }
  }
}
