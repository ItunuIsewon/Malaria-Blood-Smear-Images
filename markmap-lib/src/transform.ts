import { Remarkable } from 'remarkable';
import {
  UrlBuilder,
  JSItem,
  CSSItem,
  IMarkmapOptions,
  IMarkmapJSONOptions,
  persistJS,
  persistCSS,
  buildJSItem,
  IPureNode,
  wrapFunction,
} from 'markmap-common';
import {
  ITransformResult,
  ITransformPlugin,
  IAssets,
  ITransformHooks,
  IFeatures,
  ITransformContext,
  ITransformer,
} from './types';
import { createTransformHooks, plugins as builtInPlugins } from './plugins';
import { template, baseJsPaths } from './constants';
import { patchCSSItem, patchJSItem } from './plugins/util';

export { builtInPlugins };

function cleanNode(node: IPureNode): void {
  if (node.type === 'heading') {
    // drop all paragraphs
    node.children = node.children.filter((item) => item.type !== 'paragraph');
  } else if (node.type === 'list_item') {
    // keep first paragraph as content of list_item, drop others
    node.children = node.children.filter((item) => {
      if (['paragraph', 'fence'].includes(item.type)) {
        if (!node.content) {
          node.content = item.content;
          node.payload = {
            ...node.payload,
            ...item.payload,
          };
        }
        return false;
      }
      return true;
    });
    if (node.payload?.index != null) {
      node.content = `${node.payload.index}. ${node.content}`;
    }
  } else if (node.type === 'ordered_list') {
    let index = node.payload?.startIndex ?? 1;
    node.children.forEach((item) => {
      if (item.type === 'list_item') {
        item.payload = {
          ...item.payload,
          index,
        };
        index += 1;
      }
    });
  }
  if (node.children.length > 0) {
    node.children.forEach((child) => cleanNode(child));
    if (node.children.length === 1 && !node.children[0].content) {
      node.children = node.children[0].children;
    }
  }
}

function resetDepth(node: IPureNode, depth = 0) {
  node.depth = depth;
  node.children.forEach((child) => {
    resetDepth(child, depth + 1);
  });
}

export class Transformer implements ITransformer {
  hooks: ITransformHooks;

  md: Remarkable;

  assetsMap: Record<string, IAssets> = {};

  urlBuilder = new UrlBuilder();

  plugins: ITransformPlugin[];

  constructor(
    plugins: Array<ITransformPlugin | (() => ITransformPlugin)> = builtInPlugins
  ) {
    this.hooks = createTransformHooks(this);
    this.plugins = plugins.map((plugin) =>
      typeof plugin === 'function' ? plugin() : plugin
    );

    const assetsMap: typeof this.assetsMap = {};
    for (const { name, transform } of this.plugins) {
      assetsMap[name] = transform(this.hooks);
    }
    this.assetsMap = assetsMap;

    const md = new Remarkable('full', {
      html: true,
      breaks: true,
      maxNesting: Infinity,
    } as Remarkable.Options);
    md.renderer.rules.htmltag = wrapFunction(
      md.renderer.rules.htmltag,
      (render, ...args) => {
        const result = render(...args);
        this.hooks.htmltag.call({ args, result });
        return result;
      }
    );
    this.md = md;
    this.hooks.parser.call(md);
  }

  buildTree(tokens: Remarkable.Token[]): IPureNode {
    const { md } = this;
    const root: IPureNode = {
      type: 'root',
      depth: 0,
      content: '',
      children: [],
      payload: {},
    };
    const stack = [root];
    let depth = 0;
    for (const token of tokens) {
      const payload: IPureNode['payload'] = {};
      if (token.lines) {
        payload.lines = token.lines;
      }
      let current = stack[stack.length - 1];
      if (token.type.endsWith('_open')) {
        const type = token.type.slice(0, -5);
        if (type === 'heading') {
          depth = (token as Remarkable.HeadingOpenToken).hLevel;
          while (current?.depth >= depth) {
            stack.pop();
            current = stack[stack.length - 1];
          }
        } else {
          depth = Math.max(depth, current?.depth || 0) + 1;
          if (type === 'ordered_list') {
            payload.startIndex = (
              token as Remarkable.OrderedListOpenToken
            ).order;
          }
        }
        const item: IPureNode = {
          type,
          depth,
          payload,
          content: '',
          children: [],
        };
        current.children.push(item);
        stack.push(item);
      } else if (!current) {
        continue;
      } else if (token.type === `${current.type}_close`) {
        if (current.type === 'heading') {
          depth = current.depth;
        } else {
          stack.pop();
          depth = 0;
        }
      } else if (token.type === 'inline') {
        const revoke = this.hooks.htmltag.tap((ctx) => {
          const comment = ctx.result?.match(/^<!--([\s\S]*?)-->$/);
          const data = comment?.[1].trim().split(' ');
          if (data?.[0] === 'fold') {
            current.payload = {
              ...current.payload,
              fold: ['all', 'recursively'].includes(data[1]) ? 2 : 1,
            };
            ctx.result = '';
          }
        });
        const text = md.renderer.render([token], (md as any).options, {});
        revoke();
        current.content = `${current.content || ''}${text}`;
      } else if (token.type === 'fence') {
        const result = md.renderer.render([token], (md as any).options, {});
        current.children.push({
          type: token.type,
          depth: depth + 1,
          content: result,
          children: [],
          payload,
        });
      } else {
        // ignore other nodes
      }
    }
    return root;
  }

  transform(content: string): ITransformResult {
    const context: ITransformContext = {
      content,
      features: {},
      contentLineOffset: 0,
    };
    this.hooks.beforeParse.call(this.md, context);
    const tokens = this.md.parse(context.content, {});
    this.hooks.afterParse.call(this.md, context);
    let root = this.buildTree(tokens);
    cleanNode(root);
    if (root.children?.length === 1) root = root.children[0];
    resetDepth(root);
    return { ...context, root };
  }

  /**
   * Get all assets from enabled plugins or filter them by plugin names as keys.
   */
  getAssets(keys?: string[]): IAssets {
    const styles: CSSItem[] = [];
    const scripts: JSItem[] = [];
    keys ??= this.plugins.map((plugin) => plugin.name);
    for (const assets of keys.map((key) => this.assetsMap[key])) {
      if (assets) {
        if (assets.styles) styles.push(...assets.styles);
        if (assets.scripts) scripts.push(...assets.scripts);
      }
    }
    return {
      styles: styles.map((item) => patchCSSItem(this, item)),
      scripts: scripts.map((item) => patchJSItem(this, item)),
    };
  }

  /**
   * Get used assets by features object returned by `transform`.
   */
  getUsedAssets(features: IFeatures): IAssets {
    const keys = this.plugins
      .map((plugin) => plugin.name)
      .filter((name) => features[name]);
    return this.getAssets(keys);
  }

  fillTemplate(
    root: IPureNode | null,
    assets: IAssets,
    extra?: {
      baseJs?: JSItem[];
      jsonOptions?: IMarkmapJSONOptions;
      getOptions?: (
        jsonOptions: IMarkmapJSONOptions
      ) => Partial<IMarkmapOptions>;
    }
  ): string {
    extra = {
      ...extra,
    };
    extra.baseJs ??= baseJsPaths
      .map((path) => this.urlBuilder.getFullUrl(path))
      .map((path) => buildJSItem(path));
    const { scripts, styles } = assets;
    const cssList = [...(styles ? persistCSS(styles) : [])];
    const context = {
      getMarkmap: () => window.markmap,
      getOptions: extra.getOptions,
      jsonOptions: extra.jsonOptions,
      root,
    };
    const jsList = [
      ...persistJS(
        [
          ...extra.baseJs,
          ...(scripts || []),
          {
            type: 'iife',
            data: {
              fn: (
                getMarkmap: (typeof context)['getMarkmap'],
                getOptions: (typeof context)['getOptions'],
                root: (typeof context)['root'],
                jsonOptions: IMarkmapJSONOptions
              ) => {
                const markmap = getMarkmap();
                window.mm = markmap.Markmap.create(
                  'svg#mindmap',
                  (getOptions || markmap.deriveOptions)(jsonOptions),
                  root
                );
              },
              getParams: ({ getMarkmap, getOptions, root, jsonOptions }) => {
                return [getMarkmap, getOptions, root, jsonOptions];
              },
            },
          } as JSItem,
        ],
        context
      ),
    ];
    const html = template
      .replace('<!--CSS-->', () => cssList.join(''))
      .replace('<!--JS-->', () => jsList.join(''));
    return html;
  }
}
