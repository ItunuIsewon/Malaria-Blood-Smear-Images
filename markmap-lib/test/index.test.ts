import { Transformer } from '../src/index';

test('plugins', () => {
  const transformer = new Transformer();
  expect(transformer.plugins.map((plugin) => plugin.name)).toEqual([
    'frontmatter',
    'katex',
    'hljs',
    'npmUrl',
  ]);
  const assets = transformer.getAssets();
  expect(assets).toMatchSnapshot();
});

test('custom url provider', () => {
  const transformer = new Transformer();
  transformer.urlBuilder.setProvider('local', (path) => `/local/${path}`);
  transformer.urlBuilder.provider = 'local';
  const assets = transformer.getAssets();
  expect(assets).toMatchSnapshot();
});

test('content without frontmatter', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
- l1
  - l1.1
  - l1.2
    - l1.2.1
`);
  expect(result).toMatchSnapshot();
});

test('content with frontmatter', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
---
markmap:
  color: blue
---

- l1
  - l1.1
  - l1.2
    - l1.2.1
`);
  expect(result).toMatchSnapshot();
});

test('content with only katex enabled', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
---
markmap:
  color: blue
---

- $x = {-b \pm \sqrt{b^2-4ac} \over 2a}$
`);
  expect(result).toMatchSnapshot();
  expect(transformer.getUsedAssets(result.features)).toMatchSnapshot();
});
