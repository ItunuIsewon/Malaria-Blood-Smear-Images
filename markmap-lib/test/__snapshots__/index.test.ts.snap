// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`content with frontmatter 1`] = `
{
  "content": "
- l1
  - l1.1
  - l1.2
    - l1.2.1
",
  "contentLineOffset": 4,
  "features": {},
  "frontmatter": {
    "markmap": {
      "color": [
        "blue",
      ],
    },
  },
  "root": {
    "children": [
      {
        "children": [],
        "content": "l1.1",
        "depth": 1,
        "payload": {
          "lines": [
            2,
            3,
          ],
        },
        "type": "list_item",
      },
      {
        "children": [
          {
            "children": [],
            "content": "l1.2.1",
            "depth": 2,
            "payload": {
              "lines": [
                4,
                5,
              ],
            },
            "type": "list_item",
          },
        ],
        "content": "l1.2",
        "depth": 1,
        "payload": {
          "lines": [
            3,
            4,
          ],
        },
        "type": "list_item",
      },
    ],
    "content": "l1",
    "depth": 0,
    "payload": {
      "lines": [
        1,
        2,
      ],
    },
    "type": "list_item",
  },
}
`;

exports[`content with only katex enabled 1`] = `
{
  "content": "
- $x = {-b pm sqrt{b^2-4ac} over 2a}$
",
  "contentLineOffset": 4,
  "features": {
    "katex": true,
  },
  "frontmatter": {
    "markmap": {
      "color": [
        "blue",
      ],
    },
  },
  "root": {
    "children": [],
    "content": "<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi><mo>=</mo><mrow><mo>−</mo><mi>b</mi><mi>p</mi><mi>m</mi><mi>s</mi><mi>q</mi><mi>r</mi><mi>t</mi><mrow><msup><mi>b</mi><mn>2</mn></msup><mo>−</mo><mn>4</mn><mi>a</mi><mi>c</mi></mrow><mi>o</mi><mi>v</mi><mi>e</mi><mi>r</mi><mn>2</mn><mi>a</mi></mrow></mrow><annotation encoding="application/x-tex">x = {-b pm sqrt{b^2-4ac} over 2a}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.4306em;"></span><span class="mord mathnormal">x</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:1.0085em;vertical-align:-0.1944em;"></span><span class="mord"><span class="mord">−</span><span class="mord mathnormal">b</span><span class="mord mathnormal">p</span><span class="mord mathnormal">m</span><span class="mord mathnormal">s</span><span class="mord mathnormal" style="margin-right:0.03588em;">q</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">t</span><span class="mord"><span class="mord"><span class="mord mathnormal">b</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mord">4</span><span class="mord mathnormal">a</span><span class="mord mathnormal">c</span></span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.02778em;">er</span><span class="mord">2</span><span class="mord mathnormal">a</span></span></span></span></span>",
    "depth": 0,
    "payload": {
      "lines": [
        1,
        2,
      ],
    },
    "type": "list_item",
  },
}
`;

exports[`content with only katex enabled 2`] = `
{
  "scripts": [
    {
      "data": {
        "fn": [Function],
        "getParams": [Function],
      },
      "type": "iife",
    },
    {
      "data": {
        "defer": true,
        "src": "https://cdn.jsdelivr.net/npm/webfontloader@undefined/webfontloader.js",
      },
      "type": "script",
    },
  ],
  "styles": [
    {
      "data": {
        "href": "https://cdn.jsdelivr.net/npm/katex@undefined/dist/katex.min.css",
      },
      "type": "stylesheet",
    },
  ],
}
`;

exports[`content without frontmatter 1`] = `
{
  "content": "- l1
  - l1.1
  - l1.2
    - l1.2.1
",
  "contentLineOffset": 0,
  "features": {},
  "root": {
    "children": [
      {
        "children": [],
        "content": "l1.1",
        "depth": 1,
        "payload": {
          "lines": [
            1,
            2,
          ],
        },
        "type": "list_item",
      },
      {
        "children": [
          {
            "children": [],
            "content": "l1.2.1",
            "depth": 2,
            "payload": {
              "lines": [
                3,
                4,
              ],
            },
            "type": "list_item",
          },
        ],
        "content": "l1.2",
        "depth": 1,
        "payload": {
          "lines": [
            2,
            3,
          ],
        },
        "type": "list_item",
      },
    ],
    "content": "l1",
    "depth": 0,
    "payload": {
      "lines": [
        0,
        1,
      ],
    },
    "type": "list_item",
  },
}
`;

exports[`custom url provider 1`] = `
{
  "scripts": [
    {
      "data": {
        "fn": [Function],
        "getParams": [Function],
      },
      "type": "iife",
    },
    {
      "data": {
        "defer": true,
        "src": "/local/webfontloader@undefined/webfontloader.js",
      },
      "type": "script",
    },
  ],
  "styles": [
    {
      "data": {
        "href": "/local/katex@undefined/dist/katex.min.css",
      },
      "type": "stylesheet",
    },
    {
      "data": {
        "href": "/local/@highlightjs/cdn-assets@undefined/styles/default.min.css",
      },
      "type": "stylesheet",
    },
  ],
}
`;

exports[`plugins 1`] = `
{
  "scripts": [
    {
      "data": {
        "fn": [Function],
        "getParams": [Function],
      },
      "type": "iife",
    },
    {
      "data": {
        "defer": true,
        "src": "https://cdn.jsdelivr.net/npm/webfontloader@undefined/webfontloader.js",
      },
      "type": "script",
    },
  ],
  "styles": [
    {
      "data": {
        "href": "https://cdn.jsdelivr.net/npm/katex@undefined/dist/katex.min.css",
      },
      "type": "stylesheet",
    },
    {
      "data": {
        "href": "https://cdn.jsdelivr.net/npm/@highlightjs/cdn-assets@undefined/styles/default.min.css",
      },
      "type": "stylesheet",
    },
  ],
}
`;
